### PR TITLE
Introduce InternalChangeReplicas.

### PR DESCRIPTION
**Pull request by bdarnell**
This is the first step towards replication/rebalancing. The range state
now gets replicated correctly, although the replicated range is not yet
usable because we need to transfer metadata along with an initial
snapshot (to come in a later commit).

@spencerkimball @cockroachdb/developers

### CODE REVIEW

File: proto/api.go [hunk 1]
```diff
@@ -87,27 +87,28 @@ func (s stringSet) keys() []string {
...
-	AdminSplit:            struct{}{},
-	AdminMerge:            struct{}{},
-	Batch:                 struct{}{},
-	InternalHeartbeatTxn:  struct{}{},
-	InternalGC:            struct{}{},
-	InternalPushTxn:       struct{}{},
-	InternalResolveIntent: struct{}{},
-	InternalMerge:         struct{}{},
-	InternalTruncateLog:   struct{}{},
+	Contains:               struct{}{},
```
Comment (petermattis) [inline proto/api.go]: Doesn't `gofmt -s` get rid of the `struct{}` here?

File: storage/store.go [hunk 2]
```diff
@@ -1007,8 +1045,19 @@ func (s *Store) GroupStorage(groupID uint64) multiraft.WriteableGroupStorage {
 	defer s.mu.RUnlock()
 	r, ok := s.ranges[int64(groupID)]
 	if !ok {
-		log.Warningf("%p requested nonexistent range with raft ID %d", s, groupID)
-		return nil
+		var err error
+		r, err = NewRange(&proto.RangeDescriptor{
+			RaftID: int64(groupID),
```
Comment (petermattis) [inline storage/store.go]: Any reason we're using `int64` while etcd/raft is using `uint64` for IDs? Sort of awkward to have to cast back and forth.

**Comment by petermattis** [comment]
LGTM

File: storage/store.go (hunk 2 above)
Comment (tschottdorf) [inline storage/store.go]: Asked @spencerkimball the same thing a while ago, looked like we should be able to homogenize this but just haven't gotten around to it yet.

File: proto/api.go (hunk 1 above)
Comment (bdarnell) [inline proto/api.go]: No. The `struct{}` can be removed, but `gofmt -s` is not smart enough to do it automatically in this case.

File: multiraft/multiraft.go [hunk 3]
```diff
@@ -658,8 +679,9 @@ func (s *state) handleWriteResponse(response *writeResponse, readyGroups map[uin
 				if entry.Data != nil {
 					var command []byte
 					commandID, command = decodeCommand(entry.Data)
-					s.sendEvent(&EventCommandCommitted{commandID, command})
+					s.sendEvent(&EventCommandCommitted{groupID, commandID, command})
```
Comment (spencerkimball) [inline multiraft/multiraft.go]: Might want to set this composite literal to use more explicit initialization style.

File: multiraft/transport.go [hunk 4]
```diff
@@ -135,12 +135,24 @@ func (lt *localRPCTransport) getClient(id NodeID) (*rpc.Client, error) {
...
+	}
+
 	listener, ok := lt.listeners[id]
 	if !ok {
 		return nil, util.Errorf("unknown peer %v", id)
 	}
 	address := listener.Addr().String()
-	return rpc.Dial("tcp", address)
+
+	// If this wasn't test code we wouldn't want to call Dial while holding the lock.
```
Comment (spencerkimball) [inline multiraft/transport.go]: For these kinds of things, I usually use the format:

// NOTE: ...

File: proto/internal.pb.go [hunk 5]
```diff
@@ -333,6 +333,34 @@ func (m *InternalTruncateLogResponse) Reset()         { *m = InternalTruncateLog
 func (m *InternalTruncateLogResponse) String() string { return proto1.CompactTextString(m) }
 func (*InternalTruncateLogResponse) ProtoMessage()    {}
 
+type InternalChangeReplicasRequest struct {
```
Comment (spencerkimball) [inline proto/internal.pb.go]: By the way, is there a way to have github ignore .pb.* files in diffs?

File: proto/internal.proto [hunk 6]
```diff
@@ -161,6 +161,19 @@ message InternalTruncateLogResponse {
   optional ResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
 }
 
+message InternalChangeReplicasRequest {
+  optional RequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional int32 node_id = 2 [(gogoproto.nullable) = false,
+      (gogoproto.customname) = "NodeID", (gogoproto.customtype) = "NodeID"];
+  optional int32 store_id = 3 [(gogoproto.nullable) = false,
+      (gogoproto.customname) = "StoreID", (gogoproto.customtype) = "StoreID"];
+  optional bool remove = 4 [(gogoproto.nullable) = false];
```
Comment (spencerkimball) [inline proto/internal.proto]: Sure we don't want to make this an enum? Are we going to want to add nodes as observers, etc?

File: proto/internal.pb.go (hunk 5 above)
Comment (spencerkimball) [inline proto/internal.pb.go]: I see that github's diff already suppresses pb.cc and pb.h files...

File: storage/raft_test.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	}
+
+	args, resp := internalChangeReplicasArgs(nodeID2, storeID2, false, rangeID,
+		tc.store.StoreID())
+	if err := tc.rng.AddCmd(proto.InternalChangeReplicas, args, resp, true); err != nil {
+		t.Fatal(err)
+	}
+
+	// TODO(bdarnell): once we're hooking up the range metadata correctly,
+	// verify that the same data is on both replicas
```
Comment (spencerkimball) [inline storage/raft_test.go]: hallelujah!

File: storage/raft_test.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+	tc := testContext{}
+	tc.Start(t)
+	defer tc.Stop()
+
+	const (
+		rangeID  = 1
+		nodeID1  = 1
+		storeID1 = 1
+		nodeID2  = 2
+		storeID2 = 3
```
Comment (spencerkimball) [inline storage/raft_test.go]: Any reason stireID2 is 3, not 2?

File: proto/internal.pb.go (hunk 5 above)
Comment (bdarnell) [inline proto/internal.pb.go]: They have special-case support for recognizing c, java, and python PB files; I'll file a feature request for `.pb.go`.

https://github.com/github/linguist/pull/471

File: multiraft/multiraft.go (hunk 3 above)
Comment (bdarnell) [inline multiraft/multiraft.go]: OK.

File: multiraft/transport.go (hunk 4 above)
Comment (bdarnell) [inline multiraft/transport.go]: I understand the form `NOTE(bdarnell):` to record authorship, but what is an unadorned `NOTE:` intended to mean? How is a `NOTE:` different from a plain comment?

File: proto/internal.proto (hunk 6 above)
Comment (bdarnell) [inline proto/internal.proto]: I was hoping we'd be able to encapsulate the observer status within raft, but yeah, I'll make it an enum for future-proofing. Probably better to make a new enum here than refer to the one in `raftpb`.

File: storage/raft_test.go (hunk 8 above)
Comment (bdarnell) [inline storage/raft_test.go]: No particular reason; I was just noticing that we tend to have `nodeID == storeID == 1` in our tests and I wanted to have some cases where `nodeID != storeID`.

File: storage/store.go (hunk 2 above)
Comment (spencerkimball) [inline storage/store.go]: Yeah, I suppose we should do that refactoring. But wow, I dislike unsigned ints. I really just don't see the point. Could someone explain to me why people bother? Just because they can't be negative has always struck me as about as helpful as "const" in C++. Yes...there's utility, but you've just introduced another thing, sprinkled here and there.

There are overflow and casting issues between the two, but that has nothing to do with their broad usage. I feel like people see them as "safer" in some meaningful way which I just don't see. Is there something I'm missing?

**Comment by spencerkimball** [comment]
LGTM

File: storage/store.go (hunk 2 above)
Comment (bdarnell) [inline storage/store.go]: I agree; that's why the change hasn't happened yet (we could unilaterally switch to unsigned to match etcd/raft, but if we want to change to signed we have to convince them to change too).
